### ISSUE TITLE
Panel Buffer Makeover! [2/2]

### ISSUE THREAD

**Issue by lansingthomas** [atlas-engineer/nyxt#2984]
**Purpose**
keep fingers on keys, and off of mice while using panel buffers.

**What is your proposed change?**

in cases where the prompt buffer is closed.
- bind `Esc` to the command `delete-all-panel-buffers`

**Rationale**
the clickable exit is wonderful... I see that `X` floating there and I just want to go for a nice fast `Esc` to get rid of the panel buffer once I am done with it.

# Pre-Build Checklist:
_talk through the change with someone, then come back and mark off what you can_
- [ ] I have consulted with the team.
- [ ] One or more developers have signed off on this change.
- [ ] One or more user researchers have signed off on this change.

**Comment by aartaka** [comment]
`escape` is too generic, and we explicitly bind it to `nothing` almost everywhere, because most of us computer people hit `escape`/`C-g`(Emacs) too often.

I'd suggest something like a "close everything but this window" action for this. In Emacs, this type of action is bound to `C-x 0`. In VI bindings [it usually is `o n`](https://stackoverflow.com/questions/4809729/cancel-split-window-in-vim).

**Comment by aadcg** [comment]
I can't think of any good reason why we shouldn't bind `Esc` in CUA for this command. An alternative to it would be `C-w`.

For the Emacs keyscheme, it makes little sense to bind it `C-x 0`. It should be `C-x k`.

> because most of us computer people hit escape/C-g(Emacs) too often.

@aartaka "non-computer people" also use computers, I assure you! And they deserve attention too :)

**Comment by aartaka** [comment]
Note that `C-w` usually means killing a tab/buffer in other browsers. So it's not a perfect match.

And I still think that `escape` too weak for such a strong action as closing something.

**Comment by aadcg** [comment]
> Note that C-w usually means killing a tab/buffer in other browsers. So it's not a perfect match.

Clearly, because Nyxt doesn't match other browsers. But in the way of Nyxt, all buffers are born equal and it doesn't matter if they play the role of a web page, a side panel or others.

Regardless, it seems to me that the main topic here is that currently we don't capture whether the panel buffer has the focus, so we don't have specialized keyscheme map for it.

**Comment by aartaka** [comment]
Vivaldi uses `F4` for removing sidebar. That's the closes we can find in CUA land, so let's use it instead of more ambiguous `C-w`/`escape`.

**Comment by aadcg** [comment]
> Regardless, it seems to me that the main topic here is that currently we don't capture whether the panel buffer has the focus, so we don't provide a specialized keyscheme map for it.

That's one way of looking at it. In my mind, it would make more sense if we could kill a panel buffer in the same way that we kill a buffer. But, in our implementation, we regard panels buffers as auxiliary so we have specialized commands like `delete-panel-buffer`.

So, for now, it seems that not all buffers are born the same :(

```lisp
emacs      "C-x k" 'delete-buffer
vi-normal     "d" 'delete-buffer
```

```lisp
source/mode/base.lisp
cua       "C-w" 'delete-current-buffer
emacs      "C-x C-k" 'delete-current-buffer
vi-normal       "D" 'delete-current-buffer
source/mode/editor.lisp
cua       "C-q" 'delete-current-buffer
emacs      "C-x C-k" 'delete-current-buffer
vi-normal       "D" 'delete-current-buffer))))
source/mode/help.lisp
default       "q" 'delete-current-buffer
```

We should bind the commands related to deleting the panel buffer in a similar way to how we do it to kill a buffer.

**Comment by aartaka** [comment]
Or maybe list panels in buffer commands, maybe as a separate source :D

**Comment by aartaka** [comment]
So 
- `delete-buffer` lists panels alongside regular buffers,
- and `F4`/`C-x 0`/`o n` bound to `delete-all-panel-buffers`?

**Comment by lansingthomas** [comment]
Great ideas. :D 
Loving the back and forth.

Some thoughts....
- f4 key does not work for mac keyboards. (soon we hope to be mac compatible) 
- C-w would be nice but does not cover cases where I create a sidepanel buffer, then want to navigate to a different main buffer and keep the side panel. 
- C-q closes Nyxt so we cant use that one, people wont even experiment with it.

I think `f4` is better than nothing
I think `C-w` is better than `f4`
I think `esc` is better than `C-w`
- I think I am not following the idea that `esc` is _too weak_ to close something meaningful.  My view is that panel buffers are weaker than the main buffer so it is sensible to use `esc`. Please advise.

**Comment by jmercouris** [comment]
C-shift-w to close all panel buffers, C-w to close current buffer.

In emacs it'll be C-x 0.

Simple enough :-)

**Comment by aartaka** [comment]
> * f4 key does not work for mac keyboards. (soon we hope to be mac compatible)

Maybe Mac-specific bindings then, ahah?

> * C-w would be nice but does not cover cases where I create a sidepanel buffer, then want to navigate to a different main buffer and keep the side panel.

Not sure I get that one...

> * C-q closes Nyxt so we cant use that one, people wont even experiment with it.

Not even suggesting it!

> I think `f4` is better than nothing I think `C-w` is better than `f4` I think `esc` is better than `C-w`

- Escape is a general cancel action (leaving the input field, unfocusing the page, closing the modal), setting it to such a destructive action as deleting the panel is shooting too much feet off. That's what I meant by weakness of Escape—the regular things associated with Escape are quite benign, so one is likely to associate something reversible with it.
- `C-w` is already bound to `delete-current-buffer`. `C-s-w` is already bound to `delete-current-window`. Both of these are Nyxt-global and quite fundamental in general. Having them change their semantics depending on the context/focus is confusing, espectially until we have all buffers actually equal.
- `F4` is not portable, true. But it's there with such a semantics already. We can have it at least on *nix/Win, with Mac binding being something different (actually, lots of keyboard actions on Mac are close to Emacs, so maybe we should take a different direction altogether with regards to Mac... Another discussion sometime later!)

**Comment by aartaka** [comment]
> * f4 key does not work for mac keyboards. (soon we hope to be mac compatible)

But wait, [Magic Keyboard seems to have function keys](https://www.apple.com/shop/product/MK2A3LL/A/magic-keyboard-us-english)...

**Comment by lansingthomas** [comment]
I'm on board with f4.

well argued. 

And yeah, the magic Keyboard does actually function ( I might not use it a ton but maybe that will change once I get Nyxt on my mac machine) :)

**Comment by lansingthomas** [comment]
@jmercouris @aadcg 
`f4` seem sensible to you two?

**Comment by aadcg** [comment]
We need to be efficient in our communication. That requires thinking about a problem from top to bottom and provide a full spec. 

We need to choose keybindings for 2 commands: `delete-panel-buffer` and `delete-all-panel-buffers` (or decide that one of these commands will only be exclusively callable by name via `execute-command`). For the commands that we decide to bind, we need to think about the 4 keyscheme (default, cua, emacs and vi).

I'll suggest the spec according to the law "the lowest effort that provides maximum results".

Bind `delete-all-panel-buffers` to "F4" in the default keyscheme. Don't bind `delete-panel-buffer`.

@lansingthomas, a short clarification for you: binding in the default keyscheme means that all other keyschemes  (cua, emacs and vi) will have this keybinding available too. And don't worry, all MacOS users have access to the Function keys (F1-F12).

**Comment by jmercouris** [comment]
OK how about this then:

M-s-w for closing all panel buffers, and retain:

C-w for closing the current buffer
C-s-w for closing the current window

**Comment by aadcg** [comment]
Remarkable! I too thought deeply about alternatives to F4, and the only reasonable one was `M-w`. I kept it to myself to avoid going down the rabbit hole. After all, I try to be a peace keaper and please everyone.

That's a good spec. I prefer it over F4.

**Comment by aartaka** [comment]
F4 and Meta-prefix FTW!

**Comment by aartaka** [comment]
(However that would look strange in otherwise Meta-less CUA I guess...)

**Comment by aadcg** [comment]
@aartaka I don't understand what you're suggesting. Could you clarify?

**Comment by aartaka** [comment]
> @aartaka I don't understand what you're suggesting. Could you clarify?

I suggest F4 and `M-w` etc. The only collateral complain I have is that CUA schemes in general don't use Meta often and thus we shouldn't spruce up our default/CUA with too much Meta. But in this case it's okay, I think.

**Comment by aadcg** [comment]
@jmercouris your suggestion unfortunately won't cut it. `M-w` is bound to delete-window for CUA users. 

```
M-w for closing a specific panel buffer,
M-s-w for closing all panel buffer
```

_Efficient communication tip: try to derive all of the consequences from suggestions_ :)

-----

Then, taking into account all suggestions, I'd suggest binding the following in the default keyscheme.

```
F4 for closing a specific panel buffer,
s-F4 for closing all panel buffer
```

**Comment by lansingthomas** [comment]
Andre and I had a nice talk about this and wiggled some mockups around to see if we could do better.
The current best proposal is the one above:
```
F4 for closing a specific panel buffer,
s-F4 for closing all panel buffers
```
We identified one issue, which can be resolved.
The issue is that we are still using several keystrokes to close a simple window. 

#### possible solution is this;
1. bind `delete-panel-buffer` to f4
2. FOR instances with a single open panel buffer (and therefore a single `suggestion` in the `close-panel-buffer` UI) --automatically `return` the command without even showing the prompt buffer. 


#### Rationale for this solution:
the only option to `return` the single suggestion, so why prompt for it.

@jmercouris @aartaka sound good?

**Comment by aartaka** [comment]
> 2. FOR instances with a single open panel buffer (and therefore a single `suggestion` in the `close-panel-buffer` UI) --automatically `return` the command without even showing the prompt buffer.

This is a nice shortcut that lots of commands could benefit from. But then, having the prompt is a nice confirmation moment so that one can reconcider (irrevocable) panel deletion. So I'd rather do that later (mind opening a separate issue about that, @lansingthomas? It applies to lots of commands: `delete-buffer`, `delete-window`, `history-forwards-query`/`history-forwards` (we have two commands with two versions of this behavior right now.)) and have prompting for panel deletion in the meanwhile.

**Comment by aadcg** [comment]
I agree that this is a general issue, and @aartaka has identified it correctly. But, on the case at hand, I believe that the extra confirmation, if enabled by default, would account to a subpar experience.

**Comment by aartaka** [comment]
Oh wait, we actually have a simple solution: `prompter:auto-return-p`! Let's implement this feature and add it to all the sources requiring it in parallel.

**Comment by aadcg** [comment]
> Oh wait, we actually have a simple solution: prompter:auto-return-p!

Yes, and it makes sense that it's `nil` by default. If you're narrowing down from multiple items and arrive at a single one, it doesn't mean that you actually wanted it. Maybe you made a typo. The final confirmation makes sense.

What we're discussing here is different, I believe. It would be a `bypass-prompt-when-empty` slot that would skip prompting when the number of suggestions for all sources is 1.

> Let's implement this feature and add it to all the sources requiring it in parallel.

Could you explain?

**Comment by aartaka** [comment]
> > Let's implement this feature and add it to all the sources requiring it in parallel.
> 
> Could you explain?

I mean let's implement panel deletion, but also ensure this behavior of no-prompting-if-single-suggestion is propagated everywhere it's useful in.

**Comment by aadcg** [comment]
Understood, thanks.

I'd proceed differently. I'd close this issue by following the spec below (that we've agreed upon) and address @lansingthomas's request of ignoring prompts with a single suggestion in a separate issue. 

```
F4 for closing a specific panel buffer,
s-F4 for closing all panel buffers
```

**Comment by aartaka** [comment]
> I'd proceed differently. I'd close this issue by following the spec below (that we've agreed upon) and address @lansingthomas's request of ignoring prompts with a single suggestion in a separate issue. That request is actually an instance of a more general problem, and deserves attention by itself. We'd test the feature on this concrete use case (panel buffer deletion).

As I've suggested initially :D Let's do it this way!
